proposals/BVAL-214.md

```diff
@@ -0,0 +1,197 @@
+---
+title: Validate a list of changes without applying them to the actual object
+layout: default
+author: Gunnar Morling
+comments: true
+---
+
+# #{page.title}
+
+[Link to JIRA ticket][jira]  
+[Related JIRA](https://hibernate.atlassian.net/browse/BVAL-214)  
+
+## Problem
+
+When validating user input from a UI which is bound to the data model it is desirable to do the validation *before* propagating the values into the model.
+This prevents the model from being tainted with invalid values.
+For single properties there is `Validator#validateValue()` for this purpose, but an equivalent solution for class-level constraints is lacking today.
+
+## Proposition
+
+Provide a new method `Validator#validateValues()` similar to the existing `validateValue()` which takes several property values and validates them as if they were the actual values of the given bean types.
+
+**TL;DR:** My preference is `validateValues(Class<T> bean, ValidatedValues values, Class<?>... groups)` and exposing values through the generic property retrieval API. I like that it doesn't require cloning and enables nice cross-field validation. I think the compatability issue with existing class-level validators is acceptable, people using that feature are very likely in control of that validator implementation and can adapt it.
+
+The problem comprises two orthogonal challenges: How to pass in the values to be validated and how to expose these to constraint validators. Solutions for both are discussed below.
+
+### How to pass in the values?
+
+#### Using a map
+
+Simply a `Map<String, Object> values`.
+
+**Pro:**
+
+* Similar to `validateValue()`
+* Conceptually simple
+
+**Cons:**
+
+* Requires complete values for nested properties/collections
+* Cannot distinguish between field/property level constraints
+
+#### Using a builder API
+
+E.g. some like this:
+
+    ValidatedValues = new ValidatedValues.Builder()
+        .property( "name" ).set( "Bob ")
```

Comment:
I am undecided. I can see advantages when it comes to specifying only single properties of nested elements or single collection elements.

An alternative would be a simple expression language just for denoting bean properties:

    set( "name", "Bob" );
    set( "address['home']", new Address( ... ) );
    set( "phone.areaCode", 40 );

But devil is in the details, when it comes to conversion of map key or list index types.